### DISCUSSION TITLE
Why does paketo-libjvm set direct memory to an arbitrary 10MB by default?

### DISCUSSION THREAD

**Discussion by postalservice14** [paketo-buildpacks/feedback#241] [Java Team]
Just curious the reasoning behind setting -XX:MaxDirectMemorySize to 10M by default? The Java default seems to have always been set it to max memory. 10M as a default is quite a bit different than max memory.

**Comment by dmikusa** [comment]
I didn't make the original decision, but a lot of the Java decisions like this were made based on history and compatibility with the Cloud Foundry Java buildpack. So when you ask why, the answer is probably because that's how the CF Java Buildpack did it.

As for why the CF Java Builidpack picks this seemingly arbitrary value, I'm not sure. That predates my involvement. I do trust that the maintainers had their reasons though, likely based on research and user feedback.

Personally, I'd like to see some research into how the JVM has evolved in this area and see what we might be able to move out of the memory calculator and just let the JVM do its thing. Originally, the memory calculator was put in because it prevents container OOMEs, and the JVM had zero awareness of container limits. The JVM has been improving in this area, so maybe we can, based on the JVM version, start scaling back what the memory calculator does without risking OOMEs. I feel like when we use JVM default behaviors, that's a better user experience because it's what people naturally expect, however, preventing OOMEs is more important than that so we want to ensure we're giving users all the protection we can there. Anyway, talking out loud here.

If this is an area where you're interested in doing a little research, that would certainly help. Feel free to post findings here. Beyond that, we'd probably need an RFC to propose changes of this nature. The research I mentioned would certainly help drive that RFC.

**Reply by postalservice14** [reply]
Thanks for the quick reply @dmikusa - I'll try to summarize the why behind my question.

We have a Java/Spring project that uses Redis for session data.  The Redis client is Lettuce, which uses Netty under the hood.  Netty is know to use DirectMemory for it's IO operations.  We ran into a scenario where our application had plenty of memory, but was running out of DirectMemory and throwing an exception.  But the application wouldn't terminate because it didn't run out of memory (technically).  The result of this was strange.  The application would get the incorrect data (someone else's session).  Very bad.

Now a bit of how this has changed in the JVM.

It hasn't - lol - It looks like the JVM default has been max memory since at least Java 7.  Here are some references:

JDK 7: https://github.com/openjdk/jdk/blob/jdk7-b24/jdk/src/share/classes/sun/misc/VM.java#L192
JDK Master: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/misc/VM.java#L262

My understanding of how this works is that DirectMemory is part of the non-heap memory footprint.  So by setting it back to the java default of max memory, you could produce an OOME due to the combined DirectMemory and Heap memory exceeding the max.

For example, if you are running in a container with 2GB and you configure -Xmx2g, the DirectMemory will also get set to 2g and the total possible memory usage of the app (heap + direct) is 2 + 2 = 4GB

That doesn't sound good either, but I think the result of an OOME due to the combined heap + direct would be better than the OOME on just direct that happens today IMO.

Thanks again for your thoughts on this!

**Reply by postalservice14** [reply]
I think what would help us is even the ability to remove the setting of direct memory altogether.  I'm happy to take a stab at creating a PR to do that.

**Reply by dmikusa** [reply]
> My understanding of how this works is that DirectMemory is part of the non-heap memory footprint. So by setting it back to the java default of max memory, you could produce an OOME due to the combined DirectMemory and Heap memory exceeding the max.

Thanks for digging into this. The fact that it's unrestricted isn't something we could allow by default and it's probably not a good idea o allow users to select either. An OOME crash is really difficult to debug because in most environments/orchestrators the container and processes go away, so there's no way to go back and debug what happened. You can't get heap dumps or anything to see what was consuming that much memory. You just don't know what causes it. In my experience, they also tend to always happen at 3am when no one is around :) 

What I would suggest instead is that we look at adjusting the memory calculator's algorithm. Right now, it defaults to 10M, that's it, unless you override. Since direct memory seems to be more important to apps, we could try to do some fancier logic here. Maybe make it a fixed percent of the memory or maybe look at the app files to see if Netty or Lettuce (or some configurable list) are on the classpath, then increase the default to say 20 or 30M. Regardless of the approach, we'd still cap the direct memory it would just adjust other memory regions to compensate, i.e. more direct but less heap.

If this sounds appealing, what we'd need is a Paketo Java RFC to discuss and agree upon the new algorithm. If you look at [this RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/java/0007-low-profile-memory-calc.md) you can see an example of another RFC where we went through the process of documenting how the algorithm would work. 

Anyone can start an RFC. All you need to do is copy the one linked above, adjust it to what you propose, and submit a PR. Then we discuss. Of course, if you have questions feel free to reach out.

**Reply by postalservice14** [reply]
Yeah, I hear ya.  Thanks @dmikusa for your thoughts on this.  Let me think on this a bit and I'll open an RFC if we end up suggesting a change.

**Reply by jukkaleh-atoz** [reply]
Tl;dr: if `-XX:MaxDirectMemorySize` flag is not given [by default, the amount of native memory used for Direct Byte Buffers is limited to the maximum heap size](https://eclipse.dev/openj9/docs/xxmaxdirectmemorysize/). If the flag is given it is fixed and will OOM if full. We need an option to skip setting this flag. If larger than default size is used. Buildpack will subtract that from heap space leading to wasted ram and difficulty starting the container in azure for example. Plans pool the ram somehow and setting too large of an `XX:MaxDirectMemorySize` can lead to calculator to think there is not a enough of ram and fail to start.

We had a spring boot build with bootBuildImage gradle task. It was build to upload and download small 0-10MB pdf files to/from azure blob storage which internally uses netty to upload/download files.

This `-XX:MaxDirectMemorySize=10M` flag causes OOMs out of the blue. You can't upload/download simple files even though there is gigs of heap available.

```
java.lang.OutOfMemoryError: Cannot reserve 10282 bytes of direct buffer memory (allocated: 10478225, limit: 10485760)
   at java.base/java.nio.Bits.reserveMemory(Unknown Source) ~[na:na]
   at java.base/java.nio.DirectByteBuffer.<init>(Unknown Source) ~[na:na]
   at java.base/java.nio.ByteBuffer.allocateDirect(Unknown Source) ~[na:na]
   at io.netty.buffer.UnpooledDirectByteBuf.allocateDirect(UnpooledDirectByteBuf.java:104) ~[netty-buffer-4.1.111.Final.jar:4.1.111.Final]
```

The fix was to update the code not to load whole  file into byte array before sending it but I think we will also migrate away from buildpacks as this shouldn't have happened with the amount memory available.

From
```
public String writeFile(otherParams... final byte[] content) throws IOException {
        final Resource resource = resourceLoader.getResource(...params);
        try (final OutputStream os = ((WritableResource) resource).getOutputStream()) {
            os.write(content);
        }
        return resource.getFilename();
    }
```
Into
```
public String writeFile(otherParams... final Resource file) throws IOException {
        final Resource resource = resourceLoader.getResource(...params);
        try (final OutputStream os = ((WritableResource) resource).getOutputStream()) {
            file.getInputStream().transferTo(os);
        }
        return resource.getFilename();
    }
```

**Reply by dmikusa** [reply]
> Tl;dr: if -XX:MaxDirectMemorySize flag is not given [by default, the amount of native memory used for Direct Byte Buffers is limited to the maximum heap size](https://eclipse.dev/openj9/docs/xxmaxdirectmemorysize/).

This is exactly why we set this flag and put a limit on direct memory. The JVM is not container-friendly in this regard. If you were to go with the defaults, what would happen is that you'd wildly over-allocate memory in your container. Let's say you have a 1G limit on your container, you set the heap to 512M and non-heap except direct memory to 512M. Since you've not set direct memory to anything, the JVM will happily continue to allocate memory right up until the OOM killer steps in and kills your app.

Putting a limit on direct memory prevents this. The tradeoff is that you need to change the limit for direct memory if you need more than the default. It is very simple to override the default, just set `JAVA_TOOL_OPTIONS='-XX:MaxDirectMemorySize=20M'` or whatever you need and restart your container.

> If the flag is given it is fixed and will OOM if full. We need an option to skip setting this flag. If larger than default size is used. Buildpack will subtract that from heap space leading to wasted ram and difficulty starting the container in azure for example.

This is intentional. Again, the memory calculator does this to prevent OOMEs. If you have a 1G memory limit on your container, the memory calculator will configure the JVM to use the full 1G. If you then go and tell it to change the direct memory limit from 10M to 100M, of course, it has to change something else. That memory has to come from somewhere, otherwise, you over-allocate memory and you are now susceptible to having sporadic and unpredictable OOME's in your app.

If you need to increase direct memory and not decrease heap then you need to 1.) change the default direct memory size and 2.) increase your container's memory limit by the difference (90M in the example of going from 10M to 100M).

> This -XX:MaxDirectMemorySize=10M flag causes OOMs out of the blue. You can't upload/download simple files even though there is gigs of heap available.

This is not an issue with the buildpack or memory calculator configuration. Your app is using direct memory so it needs direct memory. It doesn't matter how much heap is available, it will fail without enough direct memory.

The reason this works for you elsewhere is simply because you are relying on the ability to over-commit your container or system's resources. That's not good practice though, because it can lead to very difficult-to-debug and painful crashes in your application. Especially in container environments because the container, and all context about the app state, goes away when the OOM killer kills the app.

The buildpack and memory calculator do not allow you to over-commit memory, it is a safety and stability feature.

**Reply by jukkaleh-atoz** [reply]
>The reason this works for you elsewhere is simply because you are relying on the ability to over-commit your container or system's resources.

Containers do have hard memory limit that is true. How you use that memory in practice should not matter, long as it doesn't run out.
If I have container with 16GB of rams I'm totally relying it not to OOM when dealing with dozen of megabytes. What I don't want is to manually joggle between direct and heap memory allocations. One day it can be 10GB of direct memory and 6GB of heap or 8gb direct and 8gb of direct memory depending on the workload. Goes totally against what Java automatic memory management is about

What I'm trying to imply is that there is no 1 fixed MaxDirectMemorySize that will work. Set it 10,20,30,40... it will always crash on some workload. Not having the flag will make the OOMs go away. Kinda nobrainer for any production workload. 

>it is a safety and stability feature.

Theoretical OOM prevention technique that actually causes OOMs

**Reply by dmikusa** [reply]
> If I have container with 16GB of rams I'm totally relying it not to OOM when dealing with dozen of megabytes. What I don't want is to manually joggle between direct and heap memory allocations. One day it can be 10GB of direct memory and 6GB of heap or 8gb direct and 8gb of direct memory depending on the workload. Goes totally against what Java automatic memory management is about

I respectfully disagree. I think that each app has its own unique memory profile. It may take a small bit of observation and tuning to dial in what that app needs, but unless an app's behavior changes drastically, the memory profile shouldn't change drastically either.

Case in point. Your app had a memory leak. You were loading an entire file into memory. That went against the expected memory profile. This setting allowed to you see that there was a problem and fix it. Your app now performs better and more memory efficient than before.

> What I'm trying to imply is that there is no 1 fixed MaxDirectMemorySize that will work. Set it 10,20,30,40... it will always crash on some workload. Not having the flag will make the OOMs go away. Kinda nobrainer for any production workload.

Again, respectfully disagree. We have this conversation come up where some user's apps are impacted by this setting, but by in large many apps work with the 10M default. Where it needs to be adjusted is when someone is either a.) doing something unique in their app with direct memory like a lot of caching or b.) there's a memory leak. This helps you detect b.) faster, which is a good thing.

**Reply by RajeevKumarMarrapu** [reply]
@dmikusa We're also facing this issue in PRD with spring boot data Redis starter which uses the lettuce client. I agree with @jukkaleh-atoz since our workloads are similar to your case

**Reply by jukkaleh-atoz** [reply]
Ultimately we migrated away from buildpacks to old schoold docker image and had zero problems afterwards which is sad because this one flag was the only known problem but absolut dealbreaker. Spring boot provides such a convenient buildpack out of the box experience. Having an option to bypass the flag would be greatly appreciated. No matter what hacky solution it would be

**Reply by maradanasai** [reply]
Are there any alternatives or workarounds? I don't want to go back to building images without packeto support in Spring Boot (gradle's bootBuildImage task) and writing Dockerfiles. Because of this issue, we did rollback for all.

How about setting -XX:MaxDirectMemorySize=0 in JAVA_TOOL_OPTIONS env? Or setting with jdk default 64mb https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/misc/VM.java#L147 ?


@dmikusa It would be nice if you can shed some light into this, thanks

**Reply by dmikusa** [reply]
@maradanasai Have you tried setting `-XX:MaxDirectMemorySize` to a larger value? You can set `-XX:MaxDirectMemorySize` yourself in `JAVA_TOOL_OPTIONS`. In fact, you can override any of the Java memory settings if that's what you want, see [instructions here](https://paketo.io/docs/howto/java/#configure-the-jvm-at-runtime). These can be changed at runtime without rebuilding your container.

The discussion above is a different topic, the OP believes they should be able to disable memory limits and while that might work for them, over the decade-plus that I've been helping users run Java apps in the cloud, the overwhelming majority of users can't make that work and will experience app crashes. It's like writing code in C, sure some subset of very experienced users can properly manage memory, but in practice, most devs will make a mistake. Buildpacks include the memory calculator to help users generate a memory configuration that will maximize the usage of the available RAM in your container without causing it to crash.

**Reply by dmikusa** [reply]
@jukkaleh-atoz I do think that at some point in the near future, we will add a way to not have the memory calculator. It is not 100% clear yet exactly what that will mean or how that will be implemented. 

The biggest challenges are a.) this is a feature for experienced users only, you're swimming at your own risk if you do this, as we don't want most users to enable this and b.) it makes things more complicated for the buildpack to enable toggling every individual feature on/off so it probably won't be fine-grained control. If you opt-out, then you're on your own, with potential memory crashes and all.

No commitments as to when this happens, but there is a small set of advanced users that want more control. If we can do that in a reasonable way, then why not.

**Reply by jukkaleh-atoz** [reply]
Nowdays cloud services at least azure doesn't allocate fixed amount of resources per container. Instead it has pool of resources included in a plan and each service in that plan share those resources dynamically. Memory calculator doesn't work great in these situation because every time you restart the service it is possible that at that memory calculator will incorrectly deduce that there isn't enough ram for direct memory + heap allocation.

Now that is a another discussion but memory calculator that thinks that resources can't scale is bit outdated thinking.

Glad to hear that it might become optional feature 👍

**Reply by dmikusa** [reply]
> Nowdays cloud services at least azure doesn't allocate fixed amount of resources per container. Instead it has pool of resources included in a plan and each service in that plan share those resources dynamically.

You'll have to explain that more to me as I don't use Azure.

**Reply by jukkaleh-atoz** [reply]
When you deploy containerized application to Azure you would probably use app service.
Each app service is assigned to [App Service plan](https://learn.microsoft.com/en-us/azure/app-service/overview-hosting-plans) and the plans are the thing that dictate the pricing and compute resources. You can switch between plans as the the applications grows/shrinks.
Having dedicated plan per app service can become expensive and is ineffecient. Azure allows/forces multiple app services to be assigned to same plan effectively sharing the resources. eg. 32 cores and 256GB of ram.

It becomes more inefficient if you have allocate gibabytes of direct memory alongside with heap allocation when they realistically could  share the same ram

Azure is in charge of allocation resources dynamically to app services per demand. So if one application is idling another can use resources instead. If all of the memory is consumed azure will restart containers until sufficient free memory is freed.

With memory calculator it checks on startup if there is enough ram. Azure will respond with some dynamic value calculated somehow. If this dynamic value is too low memory calculator won't start the container but on the next try might succeed.

**Reply by dmikusa** [reply]
Interesting. How does it handle this at the container level? If you have this pool of memory and your app needs more, what happens to the container? Does it restart it with a larger memory allocation? Does it start a new one and spin down the old one? Or is there some black magic to change the memory allocations while the container is running? If the latter, does it actually update the cgroup limits on the fly? Like if we read them later, would be different?

From what I know of the way the JVM itself handles this, it is only looking at memory limits when the JVM starts up for the first time, so if your container started with a 1G limit but was somehow adjusted to have a larger limit later, I don't think the JVM would notice that and adjust its limits. Everything I've seen says it does this when the JVM starts up, if that's changed, feel free to correct me there. If they added something to the JVM to handle adjustments that would be interesting.

Anyway, with the JVM the limits don't entirely matter because you can lie to it and tell it you have tons of memory and it won't actually care until it tries to get more memory that is actually available. So if they are doing some magic to allow you to access more memory without restarting the container, then that might keep the JVM happy, as long as it increases things fast enough that the JVM can keep getting memory.

The memory calculator is looking at the same cgroup settings that the JVM does, so it would see the same limits. Where there's a difference is that the memory calculator wants that memory upfront. It does checks to make sure it's available, with the thought being that it's better to die now rather than later when you're doing something important. If you can suddenly get access to more memory than when your app started up, that's a different thing entirely, and it invalidates the core assumptions of the memory calculator.

The core tenant of the memory calculator is how to take the memory limit for your container & get the most benefit out of that without running the risk of going over the limit, which would crash your app. That is because the limit has historically been a hard limit, where if you go over then the OOM killer steps in. If there's a soft limit, then I totally agree with you that it doesn't make sense the way it's working now. Your concerns shift at that point from trying to maximize the memory you're given in the container to how you run your app without having it consume all of your memory, which is kind of more like how it used to be pre-cloud. Where you'd put limits so that the JVM doesn't eat up all the memory on a physical box that is probably running other things too.

Very interesting. Apologizes for the million questions. This is fascinating. Thanks for sharing.

**Reply by RajeevKumarMarrapu** [reply]
> Ultimately we migrated away from buildpacks to old schoold docker image and had zero problems afterwards which is sad because this one flag was the only known problem but absolut dealbreaker. Spring boot provides such a convenient buildpack out of the box experience. Having an option to bypass the flag would be greatly appreciated. No matter what hacky solution it would be

@jukkaleh-atoz Have tried with setting default value 0 to `-XX:MaxDirectMemorySize=0`? This will let the JVM choose this value ergonomically and simplest solution, to not move from using packeto. What do you think?

**Reply by jukkaleh-atoz** [reply]
@dmikusa 
Unfortunately I don't have in-depth knoweledge of azure's internal to comment on how it works deeper. Just that they prefer selling take-or-pay computing plans.

@RajeevKumarMarrapu 
Don't know. Different jdks treat this flag differently. Not even openjdk is consistent. I was rocking jdk 17 if i remember correctly and zero didn't work on it
https://bugs.openjdk.org/browse/JDK-8284642

**Reply by maradanasai** [reply]
> Have you tried setting `-XX:MaxDirectMemorySize` to a larger value? You can set `-XX:MaxDirectMemorySize` yourself in `JAVA_TOOL_OPTIONS`. In fact, you can override any of the Java memory settings if that's what you want, see [instructions here](https://paketo.io/docs/howto/java/#configure-the-jvm-at-runtime). These can be changed at runtime without rebuilding your container.

@dmikusa Yes, we tried with this option but the problem is, I'm allocating more or less memory than needed at end. For few of my workloads, direct memory needs are not constant so I would like to handover the allocation decision to JVM instead of setting upper limit, which will reduce my heap max limit due to libjvm calc. It would be nice if we can have an option to toggle off this mem calc logic.

If this is going to take time, can you help share the steps to customise the buildpack to not include this memory calc logic and still work with bootBuildImage task in spring boot apps, thank you

**Reply by dmikusa** [reply]
> For few of my workloads, direct memory needs are not constant so I would like to handover the allocation decision to JVM instead of setting upper limit, which will reduce my heap max limit due to libjvm calc. It would be nice if we can have an option to toggle off this mem calc logic.

To be clear, the JVM does *NOT* dynamically balance this. It will set limits when it starts up too, but it's going to over-commit everything for your container limit. So if you have a memory limit on your container of 2G, it might pick a 1G heap, but it'll let you do 1G of direct memory too, and when you factor in all the other memory it needs you're over the limit for your container. This can cause crashes later if your app suddenly needs enough heap + direct + everything else to go over the container limit. 

What the JVM is doing might be a little more flexible if your application has memory needs that vary quite a bit and the JVM is able to free up heap/direct memory back to the OS so that you can share it for both. Freeing memory back to the OS is a fairly new thing for the JVM, historically the garbage collection implementations didn't do this. They just held onto memory forever. This is all the reason why the memory calculator is the way it is now.

I'm coming around to changing this, but it's not something that will happen soon. We have other projects at the moment and this isn't a small change. It will need an RFC and agreement on the path forward.

tl;dr - There's no way to overcommit memory with the memory calculator or disable the memory calculator at present. The options are to increase your memory limit and define heap/direct limits that meet the needs of your app, or not use buildpacks to package your app (or fork, you could always fork the buildpack and disable the memory calculator, but there's no configuration setting to do that).
